nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/FetchElasticsearchHttp.java

```diff
@@ -0,0 +1,312 @@
+
+                if (found) {
+                    JsonNode source = responseJson.get("_source");
+                    flowFile = session.putAttribute(flowFile, "filename", retrievedId);
+                    flowFile = session.putAttribute(flowFile, "es.index", retrievedIndex);
+                    flowFile = session.putAttribute(flowFile, "es.type", retrievedType);
+                    if (source != null) {
+                        flowFile = session.write(flowFile, out -> {
+                            out.write(source.toString().getBytes());
+                        });
+                    }
+                    logger.debug("Elasticsearch document " + retrievedId + " fetched, routing to success");
+
+                    // emit provenance event
+                    final long millis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+                    if (context.hasNonLoopConnection()) {
+                        session.getProvenanceReporter().fetch(flowFile, url.toExternalForm(), millis);
+                    } else {
+                        session.getProvenanceReporter().receive(flowFile, url.toExternalForm(), millis);
+                    }
+                    session.transfer(flowFile, REL_SUCCESS);
+                } else {
+                    logger.warn("Failed to read {}/{}/{} from Elasticsearch: Document not found",
+                            new Object[]{index, docType, docId});
+
+                    // We couldn't find the document, so send it to "not found"
+                    session.transfer(flowFile, REL_NOT_FOUND);
+                }
+            } else {
+                if (statusCode == 404) {
+                    logger.warn("Failed to read {}/{}/{} from Elasticsearch: Document not found",
+                            new Object[]{index, docType, docId});
+
+                    // We couldn't find the document, so penalize it and send it to "not found"
+                    session.transfer(flowFile, REL_NOT_FOUND);
+                } else {
+                    // 5xx -> RETRY, but a server error might last a while, so yield
+                    if (statusCode / 100 == 5) {
+
+                        logger.info("Elasticsearch returned code {} with message {}, transferring flow file to retry",
+                                new Object[]{statusCode, getResponse.message()});
+                        session.transfer(flowFile, REL_RETRY);
+                        logger.warn("Elasticsearch returned code {} with message {}, which is likely a server problem, yielding...",
+                                new Object[]{statusCode, getResponse.message()});
+                        context.yield();
+                    } else if (context.hasIncomingConnection()) {  // 1xx, 3xx, 4xx -> NO RETRY
+                        logger.info("Elasticsearch returned code {} with message {}, transferring flow file to failure", new Object[]{statusCode, getResponse.message()});
```

Comment:
Logging when sending to failure should be at least WARN (applies to line 270 as well)